src/wallet/wallet.cpp

```diff
@@ -3758,3 +3758,80 @@ void CWallet::GetFilteredNotes(
         }
     }
 }
+
+
+/* Find unspent notes filtered by payment address, min depth and max depth */
+void CWallet::GetUnspentFilteredNotes(
+    std::vector<CUnspentNotePlaintextEntry>& outEntries,
+    std::set<PaymentAddress>& filterAddresses,
+    int minDepth,
+    int maxDepth,
+    bool ignoreUnspendable)
+{
+    LOCK2(cs_main, cs_wallet);
+
+    for (auto & p : mapWallet) {
+        CWalletTx wtx = p.second;
+
+        // Filter the transactions before checking for notes
+        if (!CheckFinalTx(wtx) || wtx.GetBlocksToMaturity() > 0 || wtx.GetDepthInMainChain() < minDepth || wtx.GetDepthInMainChain() > maxDepth ) {
+            continue;
+        }
+
+        if (wtx.mapNoteData.size() == 0) {
+            continue;
+        }
+
+        for (auto & pair : wtx.mapNoteData) {
+            JSOutPoint jsop = pair.first;
+            CNoteData nd = pair.second;
+            PaymentAddress pa = nd.address;
+
+            // skip notes which belong to a different payment address in the wallet
+            if (!(filterAddresses.empty() || filterAddresses.count(pa))) {
```

Comment:
It seems like a questionable API to have the empty set of addresses mean "don't filter by address"; I would expect that not to be treated as a special case and to just return no results. ``const std::set<PaymentAddress> *`` with ``nullptr`` meaning "don't filter", or ``boost::optional<const std::set<PaymentAddress>&>`` (but note [portability issues](http://www.boost.org/doc/libs/1_62_0/libs/optional/doc/html/boost_optional/dependencies_and_portability/optional_reference_binding.html)), seem like more appropriate types.

Reply:
This approach has been used in other related methods; so beyond scope to change in this PR, but have filed ticket: https://github.com/zcash/zcash/issues/3118